### PR TITLE
Next

### PR DESCRIPTION
**Pull request by ivmarkov**

### CODE REVIEW

**Comment by mertzt89** [comment]
I've been following this branch with my project and right now the only issue I am having is that when calling .into_output() or .into_input() on a new ("disabled") PinDriver, the original PinDriver (the "disabled" one), gets dropped AFTER the pin is configured to input/output. The result is that the pin is reset to input w/ pull-up during drop() of the "disabled" PinDriver.

I've even run the blinky example and seen that this was the case.

**Comment by ivmarkov** [comment]
> I've been following this branch with my project and right now the only issue I am having is that when calling `.into_output()` or `.into_input()` on a new ("disabled") `PinDriver`, the original `PinDriver` (the "disabled" one), gets dropped AFTER the pin is configured to input/output. The result is that the pin is reset to input w/ pull-up during `drop()` of the "disabled" `PinDriver`.
> 
> I've even run the blinky example and seen that this was the case.

Good point! Let me look into this.

File: src/gpio.rs [hunk 1]
```diff
@@ -286,190 +250,1027 @@ impl From<gpio_drive_cap_t> for DriveStrength {
...
+    {
+        !self.is_low()
+    }
+
+    #[inline]
+    pub fn is_low(&self) -> bool
+    where
+        MODE: InputMode,
+    {
+        self.get_level().into()
```
Comment (mertzt89) [inline src/gpio.rs:796]: This is inverted.

**Comment by mertzt89** [comment]
> > > I've been following this branch with my project and right now the only issue I am having is that when calling `.into_output()` or `.into_input()` on a new ("disabled") `PinDriver`, the original `PinDriver` (the "disabled" one), gets dropped AFTER the pin is configured to input/output. The result is that the pin is reset to input w/ pull-up during `drop()` of the "disabled" `PinDriver`.
> > > I've even run the blinky example and seen that this was the case.
> > 
> > 
> > Good point! Let me look into this.
> 
> I think I fixed this now. Also, I've implemented `PinDriver` constructors that put the pin directly into the desired state. E.g. you can now use `PinDriver::input(...)?` instead of doing the `PinDriver::new(...)?.into_input()?` weird dance.
> 
> `PinDriver::new` is therefore now retired as well.

I tested your changes and they work correctly. Thanks!

File: src/gpio.rs (hunk 1 above)
Comment (ivmarkov) [reply src/gpio.rs]: Indeed. Fixed!

File: src/timer.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        Ok(())
+    }
+
+    pub fn enable_alarm(&mut self, enable: bool) -> Result<(), EspError> {
+        esp!(unsafe {
+            timer_set_alarm(
+                TIMER::group(),
+                TIMER::index(),
+                if enable {
+                    timer_alarm_t_TIMER_ALARM_DIS
```
Comment (usbalbin) [inline src/timer.rs:124]: Should this perhaps be
```suggestion
                    timer_alarm_t_TIMER_ALARM_EN
```

**Pull request closed without merging**